android/src/main/java/io/ably/lib/push/ActivationStateMachine.java

```diff
@@ -26,169 +26,200 @@
 
         @Override
-        public String getName() {
-            return NAME;
+        public String toString() {
+            return "GettingDeviceRegistrationFailed: " + super.toString();
         }
     }
 
     public static class GettingPushDeviceDetailsFailed extends ActivationStateMachine.ErrorEvent {
-        private static final String NAME = "GettingPushDeviceDetailsFailed";
-
         public GettingPushDeviceDetailsFailed(ErrorInfo reason) { super(reason); }
 
         @Override
-        public String getName() {
-            return NAME;
+        public String toString() {
+            return "GettingPushDeviceDetailsFailed: " + super.toString();
         }
     }
 
     public static class RegistrationSynced extends ActivationStateMachine.Event {
-        private static final String NAME = "RegistrationSynced";
+        public static final String NAME = "RegistrationSynced";
 
         @Override
-        public String getName() {
+        public String getPersistedName() {
+            return NAME;
+        }
+
+        @Override
+        public String toString() {
             return NAME;
         }
     }
 
     public static class SyncRegistrationFailed extends ActivationStateMachine.ErrorEvent {
-        private static final String NAME = "SyncRegistrationFailed";
-
         public SyncRegistrationFailed(ErrorInfo reason) { super(reason); }
 
         @Override
-        public String getName() {
-            return NAME;
+        public String toString() {
+            return "SyncRegistrationFailed: " + super.toString();
         }
     }
 
     public static class Deregistered extends ActivationStateMachine.Event {
-        private static final String NAME = "Deregistered";
+        public static final String NAME = "Deregistered";
 
         @Override
-        public String getName() {
+        public String getPersistedName() {
+            return NAME;
+        }
+
+        @Override
+        public String toString() {
             return NAME;
         }
     }
 
     public static class DeregistrationFailed extends ActivationStateMachine.ErrorEvent {
-        private static final String NAME = "DeregistrationFailed";
-
         public DeregistrationFailed(ErrorInfo reason) { super(reason); }
 
         @Override
-        public String getName() {
-            return NAME;
+        public String toString() {
+            return "DeregistrationFailed: " + super.toString();
         }
     }
 
     public abstract static class Event {
-        public abstract String getName();
-
-        public static Event constructEventByName(String className) throws ClassNotFoundException, InstantiationException {
-            ActivationStateMachine.Event event;
+        /**
+         * The name to be used when persisting this class, or null if this class should not be persisted.
+         */
+        public String getPersistedName() {
+            return null;
+        }
```

Comment:
My idea here was to force every event class to provide its own implementation of `getPersistedName()`, that why it was abstract. Main reason for that was avoid bug in the future if we add new type of event and forget to override name.

Reply:
I understand that and I recognise your point. I also appreciate very much that you are thinking in terms of future maintainability. 😍 

In this particular case, however, I believe that the balance of reduced noise in subclass implementations for what is a well established area of the codebase tips the balance in favour of providing this default implementation.